contracts/src/Utilities/ValidatorSetSig.sol

```diff
@@ -0,0 +1,137 @@
+     * cannot possibly be the source address of any other Warp message emitted by the VM.
+     */
+    address public constant VALIDATORS_SOURCE_ADDRESS = address(0);
+
+    /**
+     * @notice The blockchain ID of the chain the contract is deployed on.
+     */
+    bytes32 public immutable blockchainID;
+
+    /**
+     * @notice Warp precompile used for sending and receiving Warp messages.
+     */
+    IWarpMessenger public constant WARP_MESSENGER =
+        IWarpMessenger(0x0200000000000000000000000000000000000005);
+
+    /**
+     * @notice Emited when the payload is successfully delivered to the target contract.
+     */
+    event Delivered(address indexed targetContractAddress, uint256 indexed nonce);
+
+    constructor(bytes32 validatorBlockchainID_) {
+        validatorBlockchainID = validatorBlockchainID_;
+        blockchainID = WARP_MESSENGER.getBlockchainID();
+    }
+
+    function executeCall(uint32 messageIndex) external nonReentrant {
+        // Get the WarpMessage from the WarpMessenger pre-compile and verify that it is valid
+        (WarpMessage memory message, bool valid) =
+            WARP_MESSENGER.getVerifiedWarpMessage(messageIndex);
+        require(valid, "ValidatorSetSig: invalid warp message");
+
+        // Ensure that the sourceChainID of the validator set that signed the message is the authorized one
+        require(
+            message.sourceChainID == validatorBlockchainID, "ValidatorSetSig: invalid sourceChainID"
+        );
+        // Ensure that the originSenderAddress is the zero address to prevent on-chain messages from being forwarded
+        require(
+            message.originSenderAddress == VALIDATORS_SOURCE_ADDRESS,
+            "ValidatorSetSig: non-zero originSenderAddress"
+        );
+
+        ValidatorSetSigMessage memory validatorSetSigMessage =
+            abi.decode(message.payload, (ValidatorSetSigMessage));
+
+        validateMessage(validatorSetSigMessage);
+
+        nonces[validatorSetSigMessage.targetContractAddress] = validatorSetSigMessage.nonce + 1;
+
+        // We don't need to protect against return bomb vectors below here since the caller is expected to have full control over the contract called.
+        (bool success,) =
+        // solhint-disable-next-line avoid-low-level-calls
+         validatorSetSigMessage.targetContractAddress.call(validatorSetSigMessage.payload);
+
+        // Use require to revert the transaction if the call fails. This is to prevent consuming the nonce if the call fails due to OOG
+        // and requiring re-signing of the message with a new nonce.
+        require(success, "ValidatorSetSig: call failed");
+        emit Delivered(validatorSetSigMessage.targetContractAddress, validatorSetSigMessage.nonce);
+    }
+
+    function validateMessage(ValidatorSetSigMessage memory message) public view {
+        require(
+            message.targetBlockchainID == blockchainID,
```

Comment:
same nits here: generally have stayed away from variable name directly in require statements. Think makes sense since people using tools or submitting the txs that see the errors might not be viewing the code or have code context.